### ISSUE TITLE
Run the cross-validation for multiple iterations

### ISSUE THREAD

**Issue by JackEdTaylor** [mello-y/MSc5_research_project#7]
Splitting the sample into testing and training samples is very useful, but with such a small sample, the N=27 participants in the testing sample could differ considerably from the training sample by chance. One easy solution could be to take multiple testing and training samples (e.g., 1000 or 5000 iterations), and then concatenate the `y_pred` and `y_test` results from each iteration.

You could just put your existing code in a loop...

```python
n_iter = 5000
y_preds = []
y_tests = []

# scale before splitting into test and train samples
X_sc = StandardScaler().fit_transform(X)

for i in range(n_iter):
    # take a new testing and training sample
    X_train, X_test, y_train, y_test = train_test_split(X_sc, y, test_size = 0.25, random_state = i)
    y_tests.append(y_test)  # store the y_test sample
    # fit the logistic regression
    classifier = LogisticRegression(random_state = i, solver ='liblinear')
    classifier.fit(X_train, y_train)
    # get the y predictions and store
    y_pred = classifier.predict(X_test)
    y_preds.append(y_pred)
```

You could then generate the confusion matrix after concatenating the lists of arrays (e.g., `np.concatenate()`). If you scaled this by the sum of the array (i.e., `cm/np.sum(cm)`) then you would get probabilities for hits, misses, false positives and true negatives.

**Comment by JackEdTaylor** [comment]
Just to add to this, concatenating X_test in the same way will also let you calculate more informative ROC curves, e.g.,

![image](https://user-images.githubusercontent.com/33028594/220299937-12342468-b750-447d-ad37-b8edd65530f6.png)

**Comment by JackEdTaylor** [comment]
Just to add this, although it's a bit trickier, you could also generate a more informative ROC curve using this method. Here, you could do both the `classifier.predict()` and `classifier.predict_proba()` methods inside each iteration of the loop, and save the outputs. You would then use the `roc_auc_score()` and `roc_curve()` functions outside of the loop, as you currently do, but using the concatenated test and predictions from each iteration as arguments.

I get something like this with this method

![image](https://user-images.githubusercontent.com/33028594/220301567-8e8da67a-01ff-4429-8b96-6d1d8c84efd2.png)

**Comment by JackEdTaylor** [comment]
I see you've addressed this with a similar solution - cool! At the moment, though, your ROC curve is still just the result of the last iteration. Don't forget to use the variable name for the results from all iterations when you plot the ROC curve (i.e., `y_tests` instead of `y_test`). To do this you'll also need to alter your code for the iterations to also store and concatenate `X_test` from each iteration (e.g., in a variable called `X_tests`).

**Comment by mello-y** [comment]
When I try to do this and plot the ROC curve, it says " Found array with dim 3. Estimator expected <= 2." Idk why, since I altered the code accordingly

**Comment by JackEdTaylor** [comment]
That's interesting - sounds like one of the inputs to the ROC function has too many dimensions. Is the code pushed to any branches I can look at?

**Comment by mello-y** [comment]
Sorry Jack! Many things going on right now, I decided to let the ROV curves out since the measures should be enough. Currently working on interpreting the output and writing the discussion, and that should be it.

**Comment by JackEdTaylor** [comment]
Okay! Not sure what the problem is, but if you decide to include it and want me to have a look at the code let me know :slightly_smiling_face:
